Added-By: Gordon Malm <gengor@gentoo.org>

Note: Backported to earlier kernels. Original message included below.

---

From jejb@kernel.org  Tue Nov 11 10:17:05 2008
From: Maciej Sosnowski <maciej.sosnowski@intel.com>
Date: Tue, 11 Nov 2008 17:50:05 GMT
Subject: I/OAT: fix async_tx.callback checking
To: jejb@kernel.org, stable@kernel.org
Message-ID: <200811111750.mABHo5Ai025612@hera.kernel.org>

From: Maciej Sosnowski <maciej.sosnowski@intel.com>

commit 12ccea24e309d815d058cdc6ee8bf2c4b85f0c5f upstream

async_tx.callback should be checked for the first
not the last descriptor in the chain.

Signed-off-by: Maciej Sosnowski <maciej.sosnowski@intel.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 drivers/dma/ioat_dma.c |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

--- a/drivers/dma/ioat_dma.c
+++ b/drivers/dma/ioat_dma.c
@@ -251,7 +251,7 @@ static dma_cookie_t ioat1_tx_submit(stru
 	} while (len && (new = ioat1_dma_get_next_descriptor(ioat_chan)));
 
 	hw->ctl = IOAT_DMA_DESCRIPTOR_CTL_CP_STS;
-	if (new->async_tx.callback) {
+	if (first->async_tx.callback) {
 		hw->ctl |= IOAT_DMA_DESCRIPTOR_CTL_INT_GN;
 		if (first != new) {
 			/* move callback into to last desc */
@@ -336,7 +336,7 @@ static dma_cookie_t ioat2_tx_submit(stru
 	} while (len && (new = ioat2_dma_get_next_descriptor(ioat_chan)));
 
 	hw->ctl = IOAT_DMA_DESCRIPTOR_CTL_CP_STS;
-	if (new->async_tx.callback) {
+	if (first->async_tx.callback) {
 		hw->ctl |= IOAT_DMA_DESCRIPTOR_CTL_INT_GN;
 		if (first != new) {
 			/* move callback into to last desc */
